Kraken: Subscription improvements

gbjk opened the pull request:
* Move UpdateTradablePairs to a test
    _This should not happen in TestMain just because it's a requirement for
    most tests. It should just be a non-parallel early test.
    That allows unit tests to run quickly without calling it, and allows
    proper clear failures_
* Separate auth and non-auth ws tests
* Move SeedAssets from Setup to Start
   _Having SeedAssets in Setup is cruel and unusual because it calls the
    API. Most other interactive data seeding happens in Start.
    This made it so that fixing and creating unit tests for Kraken was
    painfully slow, particularly on flaky internet._
* Use Websocket subscriptionChannels instead of local slice
* Remove ChannelID - Deprecated in docs
   _I went both ways on this N+ times. ChannelID is great, but the bottom line is that it's deprecated, and not available in auth channels. By moving away from it we futureproof, and there's no downside really. Note that with or without_
* Simplify ping handlers and hardcodes message
* Add Depth as configurable orderbook channel param
* Simplify auth/non-auth channel updates

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-functional improvements)

## How has this been tested

- [x] go test ./... -race
- [x] golangci-lint run

gbjk commented on exchanges/kraken/kraken_test.go:
```diff
@@ -1204,61 +1210,218 @@ func TestWithdrawCancel(t *testing.T) {
+	tb.Helper()
+
+	if !k.Websocket.IsEnabled() {
+		tb.Skip("Websocket not enabled")
```
I know a different skip message has been used before.
Unless I'm missing anything, it doesn't hurt to have a more readable message here, but I didn't want to change the original.

gbjk commented on exchanges/kraken/kraken_websocket.go:
```diff
@@ -210,280 +196,257 @@ func isAwaitingCancelOrderResponses(requestID int64, success bool) bool {
+
+	event, ok := eventResponse["event"]
+	if !ok {
+		return nil
```
This is the existing behaviour before this PR.
It's not great, but I've hit MaxYakHerdOverflow
This could really do with another pass of abstraction once the dust settles after these changes, mostly to break out the sub-funcs and handle full testing of the parsing and data coming out of DataHandler channel.
I think once these immediate feature fixes are done I can take a look at those tests, or possibly @Beadko

gbjk commented on exchanges/kraken/kraken_websocket.go:
```diff
@@ -210,280 +196,257 @@ func isAwaitingCancelOrderResponses(requestID int64, success bool) bool {
+// wsPingHandler starts a websocket ping handler every 27s
+func (k *Kraken) wsPingHandler(conn stream.Connection) {
+	conn.SetupPingHandler(stream.PingHandler{
+		Message:     []byte(`{"event":"ping"}`),
```
Really didn't see the point of a fancy ping handler for this. It just feels obtuse and inefficient.

shazbert commented on exchanges/kraken/kraken_websocket.go:
```diff
@@ -210,280 +196,257 @@ func isAwaitingCancelOrderResponses(requestID int64, success bool) bool {
+		if status.Status == "error" {
+			success = false
+			cancelOrdersStatusMutex.Lock()
+			if _, ok := cancelOrdersStatus[status.RequestID]; ok {
```
Don't need to have 3 consecutive map lookups. You can name the var differently. 
```diff
diff --git a/exchanges/kraken/kraken_websocket.go b/exchanges/kraken/kraken_websocket.go
index d83c84213..5e5f6e52e 100644
--- a/exchanges/kraken/kraken_websocket.go
+++ b/exchanges/kraken/kraken_websocket.go
@@ -257,9 +257,9 @@ func (k *Kraken) wsHandleData(respRaw []byte) error {
                if status.Status == "error" {
                        success = false
                        cancelOrdersStatusMutex.Lock()
-                       if _, ok := cancelOrdersStatus[status.RequestID]; ok {
-                               if cancelOrdersStatus[status.RequestID].Error == "" { // save the first error, if any
-                                       cancelOrdersStatus[status.RequestID].Error = status.ErrorMessage
+                       if ptrToStruct, ok := cancelOrdersStatus[status.RequestID]; ok {
+                               if ptrToStruct.Error == "" { // save the first error, if any
+                                       ptrToStruct.Error = status.ErrorMessage
                                }
                        }
```

gbjk replied:
Sure. I didn't write this code at all, for what it's worth - it was just moved due to indentation fixes.

I'll fix it as a driveby though.

gbjk replied:
Nit - ptrTo is idiosyncratic go naming. 🙂

shazbert reviewed (changes requested):
Thanks for the PR. Just some quick nits and then once done I will continue with review.

gbjk commented on exchanges/kraken/kraken_websocket.go:
```diff
@@ -1222,147 +1130,216 @@ func (k *Kraken) GenerateDefaultSubscriptions() ([]stream.ChannelSubscription, e
+			}
+			switch defaultSubscribedChannels[i] {
+			case krakenWsOrderbook:
+				c.Params[ChannelOrderbookDepthKey] = krakenWsOrderbookDefaultDepth
```
Yikes! Fixed: 1abe52ef

gbjk commented on exchanges/kraken/kraken_wrapper.go:
```diff
@@ -210,11 +210,6 @@ func (k *Kraken) Setup(exch *config.Exchange) error {
 		return err
 	}
 
-	err = k.SeedAssets(context.TODO())
```
Thanks. Fixed 7e2cbc38

shazbert reviewed (approved):
LGTM thanks for making those changes. tACK. 

There are two residing issues carried over from master (out of scope). 

```
[ERROR] | ORDER | 06/10/2023 10:51:00 |  unsupported asset type
```
The is due to processing own trades. Could set order.Detail struct with Status: order.Filled to stop it trying to update it. 

Also there is noticed that there is a cycling issues when we switch from websocket to rest on orderbooks (💩 coin liquidity issues). I think I might make an update in the future so that we do not switch over if there is an active websocket connection. 

```
[ERROR] | WEBSOCKET | 06/10/2023 10:53:36 | exchange Kraken websocket error - Kraken - unhandled websocket data: Kraken XTZUSDT spot Reason: [orderbook data integrity compromised, rest sync timer lapse with active websocket connection]
```

gbjk commented:
Rebased against master.
Fixes for races and Unsub loops added.

gbjk commented:
Rebased on master and pulled in the latest changes from the bitfinex PR's websockets improvements.

gbjk commented:
@shazbert This is ready for review; Please remove the label.

Rebased on master already.

shazbert reviewed (changes requested):
Looks good:

Just some errors running instance:

```
[ERROR] | WEBSOCKET | 10/11/2023 13:02:35 | exchange Kraken websocket error - Kraken - unhandled websocket data: subscription not found: spot ohlc-1 XBT/USD
[ERROR] | WEBSOCKET | 10/11/2023 13:02:35 | exchange Kraken websocket error - Kraken - unhandled websocket data: subscription not found: spot book-1000 XBT/USD
```

gbjk commented:
> Just some errors running instance:

Ah damn it.
When I rebased on your checkSubscriptions() recently I introduced an [earlier EnsureKeyed](https://github.com/thrasher-corp/gocryptotrader/blob/2814efd7550f4177c41899713d14fd461a1e2b7e/exchanges/stream/websocket.go#L1101).
That breaks the "don't double up" check in Kraken's own `ensureKeyed`, which in hindsight wasn't very robust.

Okay. On it.

gbjk commented:
@shazbert Thanks for that.

Fixed 20d8288

shazbert reviewed (approved):
Thanks LGTM. tACK!

gbjk commented:
I'm done fiddling this PR again.
I know I circled back to fix the cancelOrder handling on this.
@shazbert Commits 52f1d68..7cdf6e9 are the new ones; Sorry to add to this after your review.

shazbert commented:
Thanks for letting me know, will check them out

gbjk commented on exchanges/kraken/kraken_test.go:
```diff
@@ -2102,3 +2128,84 @@ func TestIsPerpetualFutureCurrency(t *testing.T) {
+	}
+
+	k.API.AuthenticatedWebsocketSupport = true
+	k.SetCredentials("mock", "bW9jawo=", "", "", "", "")
```
```
SkipAuthCheck = true
```
might remove the need to do this

gbjk replied:
Fixed

shazbert commented on exchanges/kraken/kraken_test.go:
```diff
@@ -1387,11 +1377,47 @@ func TestWsAddOrder(t *testing.T) {
+				resp.Status = "error"
+				resp.ErrorMessage = "[EOrder:Unknown order]"
+			}
+			json, err := json.Marshal(resp)
```
json var -> marshalledData just so there isn't a package conflict.

gbjk replied:
Yep, agreed.
I tend to avoid using "data" for blobs of things, because it's mostly meaningless.

Fixed 05ec2ed6

shazbert commented on exchanges/kraken/kraken_test.go:
```diff
@@ -2102,3 +2116,84 @@ func TestIsPerpetualFutureCurrency(t *testing.T) {
+	return k, err
+}
+
+var upgrader = websocket.Upgrader{}
```
can rm and inline in `wsMockWrapper` below

gbjk replied:
Agreed.
I didn't think at all about why examples tended to have it as a global.

Fixed 9ee6596f

shazbert reviewed (changes requested):
LGTM one pedantic thing. Then good to go.

shazbert reviewed (approved):
Changes ack. Thanks sir.

gloriousCode commented on exchanges/kraken/kraken_types.go:
```diff
@@ -718,7 +708,7 @@ type WsCancelOrderRequest struct {
 	Event        string `json:"event"`
 	Status       string `json:"status"`
-	ErrorMessage string `json:"errorMessage"`
+	ErrorMessage string `json:"errorMessage,omitempty"`
```
This change doesn't seem necessary. The code for
`if msg, err = jsonparser.GetString(resp, "errorMessage"); err != nil {` is reading from bytes and can figure out if its missing and return. Otherwise, can you tell me how its used since I missed it?

gbjk replied:
I'm using standard json.Marshal with `WsCancelOrderResponse` in the mock handler.
This is just to make the mock responses match what the live endpoint gives :)

gloriousCode commented on exchanges/kraken/kraken_test.go:
```diff
@@ -2302,3 +2157,82 @@ func TestGetOpenInterest(t *testing.T) {
+	}
+
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { wsMockWrapper(tb, w, r, m) }))
+	if !assert.NoError(tb, err, "testInstance should not error") {
```
This looks like it can be removed

gbjk replied:
Thanks

Fixed 5919faaa

gloriousCode reviewed (changes requested):
Good stuff. Just minor things I've come across

gbjk commented on exchanges/kraken/kraken_test.go:
```diff
@@ -2302,3 +2157,82 @@ func TestGetOpenInterest(t *testing.T) {
+		}
+
+		err = m(p, c)
+		if !assert.NoError(tb, err, "WS Mock Function should not error") {
```
I don't think it should be doing the return, actually.
Errors in the mock function should not stop more processing.

I'm tending towards using assert consistently where possible for the consistency and clarity of output in the failures. I moved a few things to a t.Error() and then regretted it when I actually hit it so moved it back.
That's my reasoning, anyhow.

So i've switched it to just an NoError and I think that flies

Fixed f7a67dc8

gloriousCode replied:
I agree with the one you changed. I definitely don't want `t.Error` and `assert` used interchangably, its the readibility and the lack of t.Fatal on a test setup failure that brings my attention to the other parts. 
Good news though! testify has the additional package `require` which in my view fixes things.
```go
if strings.Contains(r.URL.Path, "GetWebSocketsToken") {
		_, err := w.Write([]byte(`{"result":{"token":"mockAuth"}}`))
		require.NoError(tb, err, "Write should not error")
		return
	}
	c, err := new(websocket.Upgrader).Upgrade(w, r, nil)
	require.NoError(tb, err, "Upgrade connection should not error")
	defer c.Close()
	for {
		_, p, err := c.ReadMessage()
		require.NoError(tb, err, "ReadMessage should not error")

		err = m(p, c)
		assert.NoError(tb, err, "WS Mock Function should not error")
	}
```
Can also help with
```go
if !assert.NoError(tb, err, "testInstance should not error") {
		tb.FailNow()
	}
```
to 
```go
require.NoError(tb, err,"testInstance should not error")
```


Uses the assert package, properly fails on a failure and makes it more readable. You can decide if these circumstances improve this specific situation and choose to use it, but this is likely how I'll be suggesting test setup failures be actioned unless there's something I've missed with it.

gbjk replied:
Perfect. I didn't spot require before, and I agree completely.

Note: This prototype of mockWS instances is being abstracted, so I'll mirror the changes out there.
I'll probably break that out into a solo PR in the coming days.

Fixed 5a66d5970

gbjk commented:
@gloriousCode Thank you 🙏 Changes made.
Let me know if you're happy and I'll fixup down before Adrian's review

gloriousCode reviewed (changes requested):
I'm happy with the functionality and testing. I would just like to avoid [a lovely black swan panicking](https://github.com/thrasher-corp/gocryptotrader/assets/9261323/bd5ea554-29f8-4ea6-ad82-b75089435832)

gbjk commented on exchanges/kraken/kraken_websocket.go:
```diff
@@ -187,303 +165,240 @@ func (k *Kraken) wsReadData(comms chan stream.Response) {
+		}
+	default:
+		k.Websocket.DataHandler <- stream.UnhandledMessageWarning{
+			Message: k.Name + stream.UnhandledMessage + string(respRaw),
```
Fixed 308789e5

Not aiming to be obtuse by switching to fmt.Sprintf, I just strongly prefer it's clarity whenever the path isn't hot-as-a-muffin 🧁

gbjk commented on exchanges/kraken/kraken_websocket.go:
```diff
@@ -1218,147 +1078,257 @@ func (k *Kraken) GenerateDefaultSubscriptions() ([]stream.ChannelSubscription, e
+}
+
+func (k *Kraken) reqForSub(e string, c *stream.ChannelSubscription) (*WebsocketSubRequest, error) {
+	r := &WebsocketSubRequest{
```
Fixed 15bc513b

gbjk commented on exchanges/kraken/kraken_websocket.go:
```diff
@@ -1218,147 +1078,257 @@ func (k *Kraken) GenerateDefaultSubscriptions() ([]stream.ChannelSubscription, e
+
+// ensureChannelKeyed wraps the channel EnsureKeyed to add channel name suffixes for Depth and Interval
+func ensureChannelKeyed(c *stream.ChannelSubscription, r *WebsocketSubRequest) error {
+	key, ok := c.EnsureKeyed().(stream.DefaultChannelKey)
```
Agreed completely. Fixed for here and all other occurrences I spotted on a quick sweep.

Fixed 15bc513b

gloriousCode reviewed (changes requested):
Thanks for those changes! Just need to fix up the lint issues from `wsProcessSpread` changes and I'm happy

gbjk commented:
@gloriousCode Build issue fixed; Ready for you

gloriousCode reviewed (approved):
tACK! Thanks for making those changes

shazbert reviewed (changes requested):
DM'd a maybe issue.

thrasher- reviewed (changes requested):
Very nice! Great improvements across the board and parallelChanOp is sexy. Tested both unauth and auth functionality. Only some minor nits

gbjk commented:
Currently re-structuring slightly `Pair` 👉 `Pairs` natively in `Subscription` and then Kraken

gbjk commented:
Will re-open this fresh. It's too cluttered and different to just force push and pretend nothing happened.
